### ISSUE TITLE
RFC on possibly changing `Author`, `License`, `CopyrightHolder`

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#1391]
__TLDR:__ I don't think any of the alternatives are better than what we have now, and suggest we keep how it's modeled now. But wanted to write something up explaining the reasoning. I'll plan on closing this issue if no substantive objections in next couple days. Also want to note that it's not a big deal to change how this is done later, so let's limit bikeshedding.

Here's current `Author`, `CopyrightHolder`, and `License` types:

```Haskell
unique type Author = Author GUID Text
unique type CopyrightHolder = CopyrightHolder GUID Text
unique type License = License [CopyrightHolder] [Year] LicenseType
```

There was some discussion about whether to change how these are modeled. A couple proposals:

```Haskell
-- 1) Get rid of the GUIDs, just use Text
unique type Author = Author Text
unique type CopyrightHolder = CopyrightHolder Text

-- 2) Get rid of the GUIDs, use an `Email` field instead for uniqueness 
unique type Author = Author Text Email
unique type CopyrightHolder = CopyrightHolder Text Email

-- 3) Orthogonal to 1 and 2, factor pull `([CopyrightHolder], [Year])` into a new type, `Copyright`
unique type Copyright = Copyright [CopyrightHolder] [Year]
unique type License = License Copyright LicenseType
```

Going through these:

Re 1) if we just have free form `Text` in there, it's unlikely people will come up with globally unique names, unless they include something globally unique like their email in that text as well (I guess using some ad hoc textual convention). If people don't use globally unique names, then that leads to unclear attribution and licensing when there's inevitable name collisions.

Re 2) The `Author` type seems fine, it's what Git does for identity - name + email. The `CopyrightHolder` type does not seem fine, as there often isn't a single email that represents a copyright holder (like a large company), and it seems weird to require that. `Email` seems better as metadata (optionally) linked to the `CopyrightHolder`. 

An alternative might be `CopyrightHolder Text [Email]`, where the email list might be empty... this might be used by a company that's a copyright holder. In general, company names are not necessarily globally unique (the full company name is probably unique within a given country, for a  period of time, but not unique globally for all time).

Re 3) Seems fine / lateral move.

Emails also aren't really globally unique, and domain names can obviously change hands. What happens when `woot.com` is shutdown, taken over by someone else, and `joe@woot.com` now means someone else entirely? Emails aren't a great primary key for long term usage.

One issue with existing codebases in other languages is that they tend to have a pretty limited set of contributors and copyright holders, so it's not a big deal if identities aren't globally unique for all time - it's generally clear from the context who "Joe Smith" is even if there are multiple. In Unison, the same definitions may be included in hundreds of libraries, so this "it's clear from context" way of disambiguating doesn't scale very well.

So basically, I don't see compelling alternatives to what's there now. Including a GUID like we are now seems like a simple and straightforward solution, we just need UCM commands to make creating `Author` and `CopyrightHolder` values easy.

**Comment by pchiusano** [comment]
Tagging @runarorama and @aryairani for review on this.

**Comment by aryairani** [comment]
> Also want to note that it's not a big deal to change how this is done later, so let's limit bikeshedding.

We can change how it's moving forward, but can't stop old metadata from being shared.  Maybe getting it wrong the first time is inevitable though, in which case it's also fine.

> Emails also aren't really globally unique, and domain names can obviously change hands. What happens when woot.com is shut down, taken over by someone else, and joe@woot.com now means someone else entirely? Emails aren't a great primary key for long term usage.

Emails are pretty globally unique though?  Maybe not capable of receiving emails forever, as domains do go down, but I personally think that the chance of joe@woot.com being a particular person now and then being a different person in the future is basically zero, and that they would both be contributing to the Unison ecosystem is much smaller.

I don't see that the GUID buys us anything in the short term over `Author "Joe Q. Functional <joe@woot.com>"` (one unstructured `Text` field w/ conventions & guidance that it be globally unique), but the downside to GUID is we are thinking it would necessitate a special command to create the values.

Is the idea that the GUID would be the primary key for the author (although anyone can currently choose non-unique values), and that they would copy that field into a new `Author` with a _new hash_ if they wanted to adjust their text field, and that if someone wanted to know if one  `Author` was the same as another, they would compare the GUIDs?

I'm not anti-GUID but I don't understand the value yet, so currently leaning Option 1).  I know we want this to be done, but if we just want quick and dirty, why not do the easy thing?  Don't wanna ignore something important though either; just help me see it.

**Comment by atacratic** [comment]
Can a definition be tagged with multiple authors?  Just wondering how it works when I edit a term that someone else originally created.

**Comment by aryairani** [comment]
@atacratic Yes; I'm thinking old and new authors could be linked to an updated definition.  I guess licensing is more complicated.

**Comment by anovstrup** [comment]
I'm wondering how code signing factors into all this, if at all. With the current scheme, _anybody_ can link _any_ author/license to _any_ definition.

What about a representation like:
```
unique type Author = { pk: PublicKey, name: Text }
unique type CopyrightHolder = { pk: PublicKey, name: Text }
unique type License = License [CopyrightHolder] [Year] LicenseType
type Signed a = Signed Signature PublicKey a
```

Authorship could be assigned by linking a definition `def` to a value `Signed signature pk h`, where `h` is the hash of `(def, author)` and `signature` is provably constructed from `sk` (the secret key corresponding to public key `pk`) and `h`. Or maybe metadata links themselves can (optionally) be signed?

**Comment by pchiusano** [comment]
@anovstrup yes, signing would be great to add later! though I would prob do so by linking the key(s) to the `Author` / `CopyrightHolder`, not by including it as a field. The current public keys being used is separate from the identity IMO.

Also noting that people use unsigned git commits quite often, and these can be spoofed as well! /cc @aryairani
